modules/fx/arrow-fx-kotlinx-coroutines/src/test/kotlin/arrow/integrations/kotlinx/CoroutinesIntegrationTest.kt

```diff
@@ -0,0 +1,166 @@
+import arrow.fx.IO
+import arrow.fx.IOOf
+import arrow.fx.extensions.fx
+import arrow.fx.extensions.io.bracket.guaranteeCase
+import arrow.fx.typeclasses.ExitCase
+import arrow.fx.typeclasses.milliseconds
+import arrow.fx.typeclasses.seconds
+import arrow.test.UnitSpec
+import arrow.test.generators.throwable
+import io.kotlintest.fail
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+import io.kotlintest.shouldBe
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.newSingleThreadContext
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.TestCoroutineExceptionHandler
+import kotlinx.coroutines.test.TestCoroutineScope
+
+@ObsoleteCoroutinesApi
+@Suppress("IMPLICIT_NOTHING_AS_TYPE_PARAMETER")
+@UseExperimental(ExperimentalCoroutinesApi::class)
+class CoroutinesIntegrationTest : UnitSpec() {
+
+  private val other = newSingleThreadContext("other")
+  private val all = newSingleThreadContext("all")
+
+  init {
+    // --------------- suspendCancellable ---------------
+
+    "suspendedCancellable should throw" {
+      forAll(Gen.throwable()) { e ->
+        val ceh = TestCoroutineExceptionHandler()
+        val scope = TestCoroutineScope(ceh + TestCoroutineDispatcher())
+
+        scope.launch {
+          IO { throw e }.suspendCancellable()
+        }
+        // suspendCancellableCoroutine re throws the exception so we need to compare the instance
+        ceh.uncaughtExceptions[0]::class == e::class
+      }
+    }
+
+    "suspendedCancellable should resume with right block" {
+      forAll(Gen.int()) { i ->
+        val ceh = TestCoroutineExceptionHandler()
+        val scope = TestCoroutineScope(ceh + TestCoroutineDispatcher())
+        scope.launch {
+          val first = IO { i + 1 }.suspendCancellable()
+          val second = IO { first + 1 }.suspendCancellable()
+          val third = IO { second + 1 }.suspendCancellable()
+
+          third shouldBe i + 3
+        }
+
+        ceh.uncaughtExceptions.isEmpty()
+      }
+    }
+
+    "scope cancellation should cancel suspendedCancellable IO" {
+      forAll(Gen.int()) { i ->
+        IO.fx {
+          val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+          val promise = !Promise<Int>()
+          !effect {
+            scope.launch {
+              IO.cancelable<Unit> { promise.complete(i) }.suspendCancellable()
+            }
+          }
+          !effect { scope.cancel() }
+          !promise.get()
+        }.unsafeRunSync() == i
+      }
+    }
+
+    "suspendCancellable can cancel even for infinite asyncs" {
+      IO.async { cb: (Either<Throwable, Int>) -> Unit ->
+        val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+        scope.launch {
+          IO.async<Int> { Thread.sleep(5000) }
```

Comment:
```suggestion
          IO.async<Int> { }
```

Shouldn't be a need to `Thread.sleep` here. `IO.never` or `IO.async<Int> { }` is the equivalent of a never finishing async task.